go-controller/pkg/ovn/egressgw.go

```diff
@@ -951,6 +951,19 @@ func (oc *Controller) cleanExGwECMPRoutes() {
 		klog.Infof("Syncing exgw routes took %v", time.Since(start))
 	}()
 
+	// migration from LGW to SGW mode
```

Comment:
It might be my own unfamiliarity with this egress gateway code, but why does moving the same bloc of code ~100 lines up (with no return statement in between those line, bar the `ovnRouteCache` check) change anything? I assume the `ovnRouteCache` check to not matter because, as the commented code says: there's nothing in OVN so nothing to do?

Reply:
the ovnRouteCache checks to see if there are any ECMP routes in OVN on the GRs. If there are none, we assume we dont need to go through and sync anything. That is normally correct, because programming ECMP routes is typed to programming logical route policies (LRP) in ovn_cluster_router. In other words, to configure an ECMP route for local gw mode, we would have also programmed an LRP. So basically if we see the ecmp routes are 0 (ovnRouteCache) we assume there is nothing that needs to be done.

This is probably not a 100% accurate assumption and we should also go remove all 501 policies len of ovnRoutecache is 0. I can add that. However, besides this issue when migrating from shared -> local or when cleaning up legacy policies we should do that no matter if there are ECMP routes or not, which is why I moved it to the top. Does that make sense?

Example: someone created a single exgw pod in local gw mode, and then deleted it. But for some reason the delete succeeded on the ECMP route and failed on the LRP delete. Next time ovnkube-master restarts and syncs it would miss removing that LRP with the way the current code is. If the user upgraded to newer OVN-K or switched to shared gw mode it would also miss cleaning it up.